eclass/databases.eclass

```diff
@@ -0,0 +1,505 @@
+#
+# --get-depend
+#
+# 	Returns a dependency string (to be included in BDEPEND).
+#
+# --get-logfile
+#
+# 	Returns the path to the server's log file.
+#
+# --get-pidfile
+#
+# 	Returns the path to the server's PID file.
+#
+# --get-sockdir
+#
+# 	Returns the directory where the server's sockets are located.
+#
+# --get-sockfile
+#
+# 	Returns the path to the server's socket file.
+#
+# --start
+#
+# 	Starts the server on the default port.
+#
+# --start <port>
+#
+# 	Starts the server on the given port.
+#
+# --stop
+#
+# 	Stops the server.
+#
+# @EXAMPLE:
+#
+# @CODE
+# EAPI=8
+#
+# ...
+#
+# inherit databases distutils-r1
+#
+# ...
+#
+# BDEPEND="$(eredis --get-depend)"
+#
+# distutils_enabe_tests pytest
+#
+# src_test() {
+# 	eredis --start 16739
+# 	distutils-r1_src_test
+# 	eredis --stop
+# }
+# @CODE
+
+case ${EAPI} in
+	8) ;;
+	*) die "${ECLASS}: EAPI ${EAPI} unsupported."
+esac
+
+if [[ ! ${_DATABASES_ECLASS} ]]; then
+_DATABASES_ECLASS=1
+
+# ==============================================================================
+# GENERIC FUNCTIONS
+# ==============================================================================
+
+# @FUNCTION: _databases_gen_depend
+# @USAGE: <funcname>
+# @INTERNAL
+# @DESCRIPTION:
+# Get a dependency string for the given helper function.
+_databases_gen_depend() {
+	local srvname=${1:1}
+	case ${srvname} in
+		memcached)
+			echo "net-misc/memcached"
+			;;
+		mongod)
+			echo "dev-db/mongodb"
+			;;
+		mysql)
+			echo "virtual/mysql[server]"
+			;;
+		postgres)
+			echo "dev-db/postgresql[server]"
+			;;
+		redis)
+			echo "dev-db/redis"
+			;;
+		*)
+			die "${ECLASS}: unknown database: ${srvname}"
+	esac
+}
+
+# @FUNCTION: _databases_die
+# @USAGE: <funcname> [msg]
+# @INTERNAL
+# @DESCRIPTION:
+# Print the given message and the path to the server's log file to the console
+# and die.
+#
+# This function supports being called via "nonfatal".
+_databases_die() {
+	local funcname=${1?}
+	shift
+
+	eerror "See the server log for details:"
+	eerror "	$(${funcname} --get-logfile)"
+	die -n "${@}"
+}
+
+# @FUNCTION: _databases_mkdirs
+# @USAGE: <funcname>
+# @INTERNAL
+# @DESCRIPTION:
+# Create all directories that might be needed at runtime.
+_databases_mkdirs() {
+	debug-print-function "${FUNCNAME}" "${@}"
+
+	local funcname=${1?}
+
+	# directories
+	local dbpath=$(${funcname} --get-dbpath)
+	local sockdir=$(${funcname} --get-sockdir)
+	mkdir -p "${dbpath}" || die "Creating database directory failed"
+	mkdir -p "${sockdir}" || die "Creating sockets directory failed"
+
+	# files
+	local logfile_dir=$(dirname "$(${funcname} --get-logfile)")
+	local pidfile_dir=$(dirname "$(${funcname} --get-pidfile)")
+	local sockfile_dir=$(dirname "$(${funcname} --get-sockfile)")
+	mkdir -p "${logfile_dir}" || die "Creating log directory failed"
+	mkdir -p "${pidfile_dir}" || die "Creating pidfile directory filed"
+	mkdir -p "${sockfile_dir}" || die "Creating sockets directory filed"
```

Comment:
You can collapse these into a single`mkdir` now, since they're all using the same directory, and avoid `dirname` in the process.

Reply:
Done